File: architecture-skills/design-patterns/examples/strategy/strategy.php
<?php
// Strategy pattern in PHP.
//
// PHP closures and first-class callables cover stateless strategies
// (usort's comparator is a strategy). An interface earns its keep when
// the strategy carries state or several methods; both forms are shown.

declare(strict_types=1);

final class Order
{
    public function __construct(
        public readonly float $weightKg,
        public readonly float $distanceKm,
    ) {
    }
}

interface ShippingStrategy
{
    public function cost(Order $order): float;
}

// A stateful strategy: the negotiated rates are its state.
final class CarrierRate implements ShippingStrategy
{
    public function __construct(
        private readonly float $base,
        private readonly float $perKg,
    ) {
    }

    public function cost(Order $order): float
    {
        return $this->base + $this->perKg * $order->weightKg;
    }
}

// Context: accepts a closure or an object; delegates pricing to it.
final class Checkout
{
    /** @var callable(Order): float */
    private $strategy;

    public function __construct(callable $strategy)
    {
        $this->strategy = $strategy;
    }

    public function setStrategy(callable $strategy): void
    {
        $this->strategy = $strategy;
    }

    public function shippingFor(Order $order): float
    {
        return ($this->strategy)($order);
    }
}

$order = new Order(weightKg: 2.4, distanceKm: 120);

$strategies = [
    'flat' => fn (Order $o): float => 5.00,
    'weight' => fn (Order $o): float => 4.00 + 1.50 * $o->weightKg,
    'distance' => fn (Order $o): float => 2.00 + 0.05 * $o->distanceKm,
    'carrier' => (new CarrierRate(base: 3.00, perKg: 2.00))->cost(...),
];

$checkout = new Checkout($strategies['flat']);
foreach ($strategies as $name => $strategy) {
    $checkout->setStrategy($strategy);
    printf("%8s: $%.2f\n", $name, $checkout->shippingFor($order));
}

File: architecture-skills/design-patterns/examples/strategy/strategy.rs
// Strategy pattern in Rust.
//
// Stateless strategies are idiomatically just closures or fn pointers
// (compare sort_by taking a comparator), so the context stores a boxed
// Fn. A trait is the right tool when a strategy carries state or
// several methods; the trait form plugs into the same context through
// a closure that captures the strategy value.

struct Order {
    weight_kg: f64,
    distance_km: f64,
}

// --- Closure form: idiomatic for stateless strategies ---

type ShippingStrategy = Box<dyn Fn(&Order) -> f64>;

/// Context: delegates the pricing decision to the current strategy.
struct Checkout {
    strategy: ShippingStrategy,
}

impl Checkout {
    fn new(strategy: ShippingStrategy) -> Self {
        Self { strategy }
    }

    fn set_strategy(&mut self, strategy: ShippingStrategy) {
        self.strategy = strategy;
    }

    fn shipping_for(&self, order: &Order) -> f64 {
        (self.strategy)(order)
    }
}

// --- Trait form: for stateful or multi-method strategies ---

trait ShippingCost {
    fn cost(&self, order: &Order) -> f64;
}

/// A stateful strategy: the negotiated rates are its state.
struct CarrierRate {
    base: f64,
    per_kg: f64,
}

impl ShippingCost for CarrierRate {
    fn cost(&self, order: &Order) -> f64 {
        self.base + self.per_kg * order.weight_kg
    }
}

fn main() {
    let order = Order {
        weight_kg: 2.4,
        distance_km: 120.0,
    };

    let carrier = CarrierRate {
        base: 3.00,
        per_kg: 2.00,
    };

    let strategies: Vec<(&str, ShippingStrategy)> = vec![
        ("flat", Box::new(|_: &Order| 5.00)),
        ("weight", Box::new(|o: &Order| 4.00 + 1.50 * o.weight_kg)),
        ("distance", Box::new(|o: &Order| 2.00 + 0.05 * o.distance_km)),
        ("carrier", Box::new(move |o: &Order| carrier.cost(o))),
    ];

    let mut checkout = Checkout::new(Box::new(|_: &Order| 5.00));
    for (name, strategy) in strategies {
        checkout.set_strategy(strategy);
        println!("{name:>8}: ${:.2}", checkout.shipping_for(&order));
    }
}
